AudioKit/Common/Nodes/Playback/Player/AKPlayer.swift

```diff
@@ -0,0 +1,724 @@
+
+        } catch let err as NSError {
+            AKLog("ERROR AKPlayer: Couldn't read data into buffer. \(err)")
+            return
+        }
+
+        // Now, we'll reverse the data in the buffer if specified
+        if isReversed {
+            reverseBuffer()
+        }
+
+        // Same idea with fades
+        if isFaded {
+            fadeBuffer()
+            fade.needsUpdate = false
+        }
+
+        // these are only stored to check if the buffer needs to be updated in subsequent fills
+        self.startingFrame = startFrame
+        self.endingFrame = endFrame
+
+        Swift.print("updateBuffer() read from \(startTime) to \(endTime)")
+    }
+
+    // Apply sample level fades to the internal buffer.
+    // TODO: add other fade curves or ditch this method in favor of Audio Unit based fading.
+    // That is appealing as it will work with file playback as well as buffer
+    private func fadeBuffer() {
+        if fade.inTime == 0 && fade.outTime == 0 {
+            return
+        }
+        AKLog("fadeBuffer() inTime: \(fade.inTime) outTime: \(fade.outTime)")
+
+        guard isBuffered, let buffer = self.buffer, let audioFile = self.audioFile else { return }
+
+        let fadedBuffer = AVAudioPCMBuffer(pcmFormat: buffer.format, frameCapacity: buffer.frameCapacity)
+        let length: AVAudioFrameCount = buffer.frameLength
+
+        // initial starting point for the gain, if there is a fade in, start it at .01 otherwise at 1
+        var gain: Double = fade.inTime > 0 ? fade.inStartGain : 1.0
+
+        let sampleTime: Double = 1.0 / audioFile.sampleRate
+
+        //exponential fade type
+
+        // from -20db?
+        let fadeInPower: Double = exp(log(10) * sampleTime / fade.inTime)
+
+        // for decay to x% amplitude (-dB) over the given decay time
+        let fadeOutPower: Double = exp(-log(25) * sampleTime / fade.outTime)
+
+        // where in the buffer to end the fade in
+        let fadeInSamples = Int(audioFile.sampleRate * fade.inTime)
+        // where in the buffer to start the fade out
+        let fadeOutSamples = Int(Double(length) - (audioFile.sampleRate * fade.outTime))
+
+        // i is the index in the buffer
+        for i in 0 ..< Int(length) {
+            // n is the channel
+            for n in 0 ..< Int(buffer.format.channelCount) {
+
+                if i <= fadeInSamples && fade.inTime > 0 {
+                    gain *= fadeInPower
+                } else if i >= fadeOutSamples && fade.outTime > 0 {
+                    if i == fadeOutSamples {
+                        gain = fade.outStartGain
+                    }
+                    gain *= fadeOutPower
+                } else {
+                    gain = 1.0
+                }
+
+                //sanity check
+                if gain > 1 {
+                    gain = 1
+                }
+
+                let sample = buffer.floatChannelData![n][i] * Float(gain)
```

Comment:
Unwrapped optional again.
Maybe throw this in with your guard statement at the top.

```
guard isBuffered,
    let buffer = self.buffer,
    let audioFile = self.audioFile,
    let floatChannelData = buffer.floatChannelData else {
        return
}
```

Reply:
Might as well through fadedBuffer in the guard statement too as it's integral to what the function does.